### PR TITLE
NEXT-12713 - [NEW] Override existing routes

### PR DESCRIPTION
**Pull request by NiklasLimberg**
## What should be done?

Create a new article on our [GitBook instance|https://app.gitbook.com/@shopware/s/shopware/] which explains how to override existing storefront routes.

This article should mention:
- The prerequisite, a working plugin (Refer to the plugin base guide)
- Every other prerequisite you figure out during writing the guide (e.g. a subscriber, knowing how to create a service, a controller, etc.)
- A short code example, including an explanation, on how to do it

Category: Extensions > Plugins > Storefront

## Are there already guides in the previous documentation for this?
Unfortunately, No. But please have a look at our [example guide|https://app.gitbook.com/@shopware/s/shopware/guides/plugins/plugins/plugin-base-guide](e.g. note the PLACEHOLDERs for cross-references, that do not exist yet) and the [Writing guidelines|https://app.gitbook.com/@shopware/s/shopware/resources/guidelines/documentation/writing].

For more information, ask me, [~p.stahl].

### CODE REVIEW

**Review by raknison — APPROVED** [review]

File: guides/plugins/plugins/administration/customizing-modules.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ A `module` is an encapsulated unit which implements a whole feature. For example
 
 ## Prerequisites
 
-All you need for this guide is a running Shopware 6 instance. Of course, you'll have to understand JavaScript and have a basic familiarity with TwigJS, the templating engine, used in the Administration. However, that's a prerequisite for Shopware as a whole and will not be taught as part of this documentation.
+All you need for this guide is a running Shopware 6 instance. Of course, you'll have to understand JavaScript and have a basic familiarity with [Vue](https://vuejs.org/) and the [Vue Router](https://router.vuejs.org/). However, that's a prerequisite for Shopware as a whole and will not be taught as part of this documentation.
```
Comment (lernhart) [inline guides/plugins/plugins/administration/customizing-modules.md:10]:
```suggestion
All you need for this guide is a running Shopware 6 instance. Of course, you'll have to understand JavaScript and have a basic familiarity with [Vue](https://vuejs.org/) and the [Vue Router](https://router.vuejs.org/). However, these are prerequisites for Shopware as a whole and will not be taught as part of this documentation.
```

File: guides/plugins/plugins/administration/overriding-routes.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# Overriding routes
+
+## Overview
+
+In the `Administration` core code, each module is defined in a directory called `module`. Modules define routes wich can be extended with `routeMiddleware`.
+Learn what else you can customize, in existing modules in this [guide](./customizing-modules.md)
```
Comment (lernhart) [inline guides/plugins/plugins/administration/overriding-routes.md:6]:
```suggestion
To see what else you can customize in existing modules, have a look at this [guide](./customizing-modules.md)
```

File: guides/plugins/plugins/administration/overriding-routes.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+A `module` is an encapsulated unit which implements a whole feature. For example there are modules for customers, orders, settings, etc.
+
+## Prerequisites
```
Comment (lernhart) [inline guides/plugins/plugins/administration/overriding-routes.md:10]: Aren't we having a basic modules guide as a prerequisite here ?

File: guides/plugins/plugins/administration/overriding-routes.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+All you need for this guide is a running Shopware 6 instance. Of course, you'll have to understand JavaScript and have a basic familiarity with [Vue](https://vuejs.org/) and the [Vue Router](https://router.vuejs.org/). However, that's a prerequisite for Shopware as a whole and will not be taught as part of this documentation.
+
+## Appling the override
+
+At some point you might want to override or change existing routes, to for example, change the privileges required for a route or entirely replace it with your own.
+
+This is done by creating a new module and implementing a `routeMiddleware`.
```
Comment (lernhart) [inline guides/plugins/plugins/administration/overriding-routes.md:18]:
```suggestion
This is done by creating a new module and implementing a `routeMiddleware` method.
```
Can this bew called a  `method` ?

**Review by lernhart — CHANGES_REQUESTED** [review]

File: guides/plugins/plugins/administration/overriding-routes.md (hunk 4 above)
Comment (NiklasLimberg) [reply guides/plugins/plugins/administration/overriding-routes.md:18]: it's a middleware method, so technically jes, but i would leave it as it is

File: guides/plugins/plugins/administration/overriding-routes.md (hunk 3 above)
Comment (NiklasLimberg) [reply guides/plugins/plugins/administration/overriding-routes.md:10]: probably a good idea

File: guides/plugins/plugins/administration/overriding-routes.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# Overriding routes
+
+## Overview
+
+In the `Administration` core code, each module is defined in a directory called `module`. Modules define routes wich can be extended with `routeMiddleware`.
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/overriding-routes.md:5]:
```suggestion
In the `Administration` core code, each module is defined in a directory called `module`. Modules define routes which can be extended with `routeMiddleware`.
```

File: guides/plugins/plugins/administration/overriding-routes.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+To see what else you can customize in existing modules, have a look at this [guide](./customizing-modules.md)
+
+A `module` is an encapsulated unit which implements a whole feature. For example there are modules for customers, orders, settings, etc.
+
+## Prerequisites
+
+All you need for this guide is a running Shopware 6 instance. Of course, you'll have to understand JavaScript and have a basic familiarity with [Vue](https://vuejs.org/) and the [Vue Router](https://router.vuejs.org/). However, that's a prerequisite for Shopware as a whole and will not be taught as part of this documentation. 
+Further a basic understanding of what modules are is also required, learn more about them [here](./add-custom-module.md) 
+
+## Appling the override
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/overriding-routes.md:15]:
```suggestion
## Applying the override
```

File: guides/plugins/plugins/administration/overriding-routes.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+At some point you might want to override or change existing routes, to for example, change the privileges required for a route or entirely replace it with your own.
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/overriding-routes.md:17]:
```suggestion
At some point you might want to override or change existing routes, for example, to change the privileges required for a route or entirely replace it with your own.
```

File: guides/plugins/plugins/administration/overriding-routes.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        next(currentRoute);
+    }
+});
+```
+{% endcode %}
+
+This `routeMiddleware` changes the required privileges for the `sw.product.detail.base` route from `product.viewer` to `product.editor`.
+The rest of the route configurations stays the same in this example.
+
+If you want to learn more about ACL take a look at this [guide](./add-acl-rules.md) and if you want to learn everything about administration routes, take a look at this [guide](./add-custom-route.md)
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/overriding-routes.md:49]:
```suggestion
If you want to learn more about ACL take a look at this [guide](./add-acl-rules.md) and if you want to learn everything about administration routes, head over to this [guide](./add-custom-route.md)
```

**Review by leichteckig — CHANGES_REQUESTED** [review]

**Review by leichteckig — APPROVED** [review]

File: guides/plugins/plugins/administration/overriding-routes.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+});
+```
+{% endcode %}
+
+This `routeMiddleware` changes the required privileges for the `sw.product.detail.base` route from `product.viewer` to `product.editor`.
+The rest of the route configurations stays the same in this example.
+
+If you want to learn more about ACL take a look at this [guide](./add-acl-rules.md) and if you want to learn everything about administration routes, head over to this [guide](./add-custom-route.md)
+
+## Next steps
```
Comment (PaddyS) [inline guides/plugins/plugins/administration/overriding-routes.md:51]: Can be removed

**Review by PaddyS — CHANGES_REQUESTED** [review]

**Comment by PaddyS** [comment]
Also squash & rebase please

**Pull request merged**